### ISSUE TITLE
devcontainers: cannot build container and establish connection with it after upgrading to VSCode versions 1.87.2 and later

### ISSUE THREAD

**Issue by astrotarkus** [microsoft/vscode-remote-release#9920]
- VSCode Version: 1.89.1, problem first appeared in 1.87.2
- Local OS Version: Windows 11 with repository inside WSL2 Ubuntu 24.04
- Remote OS Version: Ubuntu 24.04 also running in docker container
- Remote Extension/Connection Type: Containers

My colleagues and I can no longer build our docker container from within VSCode and connect to it after upgrading VSCode. The last version which behaved as expected (container launched and connection successful) was 1.86.2. Now running 1.89.1, but the problem seems to have first appeared in 1.87.2 or 1.87.1. The devcontainers extension version remains the same at 0.362.0 release version. 

- Logs:
[remoteContainers-2024-05-24T01-05-00.792Z_1.89.1_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15426398/remoteContainers-2024-05-24T01-05-00.792Z_1.89.1_notworking.log)
[remoteContainers-2024-05-24T01-08-53.826Z_1.86.2_working.log](https://github.com/microsoft/vscode-remote-release/files/15426399/remoteContainers-2024-05-24T01-08-53.826Z_1.86.2_working.log)
[renderer_1.86.2_working.log](https://github.com/microsoft/vscode-remote-release/files/15426400/renderer_1.86.2_working.log)
[renderer_1.89.1_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15426401/renderer_1.89.1_notworking.log)
[shutdownMonitor-2024-05-24T01-03-03.605Z_1.89.1_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15426402/shutdownMonitor-2024-05-24T01-03-03.605Z_1.89.1_notworking.log)
[shutdownMonitor-2024-05-24T01-08-53.826Z_1.86.2_working.log](https://github.com/microsoft/vscode-remote-release/files/15426404/shutdownMonitor-2024-05-24T01-08-53.826Z_1.86.2_working.log)

Let me know if more logs are needed.

Steps to Reproduce:
1. Open workspace in VSCode
2. Open container menu and select 'Reopen in container'
An error message pops up stating "devcontainers: command failed {0}", and the container remains disconnected. Looking at the logs, it seems the container just doesn't respond after an attempt to start it. 

Do note that if the container is already built, 'Reopen in container' succeeds in post 1.86.2 versions. So it appears it can start it if it is already built, but cannot build it. 'Rebuild Container' fails in the affected versions. 

I have tried the pre-release 0.366.0 devcontainers version but no dice.

**Comment by tanyo520** [comment]
+1

**Comment by chrmarti** [comment]
@astrotarkus Could you check if it is fixed in Insiders? https://code.visualstudio.com/insiders/

@bpasero I wanted to suggest using the command below, but that errors out claiming that is not a released Insiders build. I picked the last released 1.86.0 Insiders. Any idea?
```
npx @vscode/vscode-bisect --good 8c08612b658a37f96892f8ed5a9c6e7a6667847d
```

**Comment by bpasero** [comment]
@chrmarti what does `vscode-bisect --help` print as version?

**Comment by chrmarti** [comment]
@bpasero Version 0.4.6

**Comment by bpasero** [comment]
@chrmarti we are only able to fetch the last 200 builds and this one is likely before. Thats a hard limit on the update server that would need a change there.

**Comment by chrmarti** [comment]
@bpasero The query includes _unreleased_ Insiders builds, when I only list released Insiders builds, the build in question is among those listed.

When including unreleased builds the oldest build listed is 4 months old, when only including released builds the oldest build is 7 months old.

Not sure if you need unreleased builds there: https://github.com/microsoft/vscode-bisect/blame/8b418cc84532a7aeb8770459820929f3465f23a1/src/builds.ts#L73

**Comment by bpasero** [comment]
@chrmarti yeah thanks for the idea, I am adding a new option `--releasedOnly` that allows you to bisect only over released builds, giving you that larger time window. See https://github.com/microsoft/vscode-bisect/pull/11

**Comment by chrmarti** [comment]
@bpasero Looks good!

@astrotarkus If VS Code Insiders doesn't fix it, could you run the following to find the Insiders version that broke it:
```
npx @vscode/vscode-bisect --good 8c08612b658a37f96892f8ed5a9c6e7a6667847d --releasedOnly
```
This will download and start a series of Insiders builds and ask you if the problem exists in that version.

**Comment by astrotarkus** [comment]
@chrmarti Thanks for picking this up. Tried with Insiders 1.90, unfortunately this doesn't fix it. Here are the logs: 
[remoteContainers-2024-05-29T03-35-37.673Z_1.90_insiders_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15478453/remoteContainers-2024-05-29T03-35-37.673Z_1.90_insiders_notworking.log)
[renderer_1.90_insiders_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15478454/renderer_1.90_insiders_notworking.log)
[shutdownMonitor-2024-05-29T03-35-20.544Z_1.90_insiders_notworking.log](https://github.com/microsoft/vscode-remote-release/files/15478455/shutdownMonitor-2024-05-29T03-35-20.544Z_1.90_insiders_notworking.log)

In the process of finding the build that broke it, will get back to you shortly.

**Comment by astrotarkus** [comment]
Unfortunately I'm having trouble installing npm due to our company firewall - is there a way I can do this manually? I do know it was broken between the released versions 1.86.2 and 1.87.2 - can narrow it down to .0 or .1

**Comment by chrmarti** [comment]
There are a lot of changes between 1.86 and 1.87. The bisect tool would narrow it down to a single Insiders version which often only has a single day of changes compared to the previous Insiders version.

The log suggests that the Dev Containers CLI tried to start the container but then the CLI stopped executing without further error message.

@bpasero This might be unrelated: The renderer log shows several file watcher errors.

**Comment by astrotarkus** [comment]
Managed to get the bisect tool working, and it tells me that ec291c126878742ad640055ce604a58129cd088c is the first bad commit. 
https://github.com/microsoft/vscode/compare/2af613979f646fc4dcebfeaedc7d14f138c7b072...ec291c126878742ad640055ce604a58129cd088c

**Comment by chrmarti** [comment]
@deepak1556 This seems to coincide with the update of node-pty to use NAPI (though I can't find the corresponding tags in the node-pty repository). The log ends at the point where we launch `docker run` using node-pty.

**Comment by astrotarkus** [comment]
[vscode log 1.86.2 - working.zip](https://github.com/user-attachments/files/15541215/vscode.log.1.86.2.-.working.zip)
[vscode log 1.89.1 - not working.zip](https://github.com/user-attachments/files/15541217/vscode.log.1.89.1.-.not.working.zip)
 Here are zips of all the log files if that is any help.
